File: mifosng-android/src/main/java/com/mifos/mifosxdroid/activity/pathtracking/PathTrackingPresenter.java
package com.mifos.mifosxdroid.activity.pathtracking;

import com.mifos.api.datamanager.DataManagerDataTable;
import com.mifos.mifosxdroid.base.BasePresenter;
import com.mifos.objects.user.UserLocation;

import java.util.List;

import javax.inject.Inject;

import rx.Subscriber;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;
import rx.subscriptions.CompositeSubscription;

/**
 * Created by Rajan Maurya on 24/01/17.
 */

public class PathTrackingPresenter extends BasePresenter<PathTrackingMvpView> {

    private final DataManagerDataTable dataManagerDataTable;
    private CompositeSubscription subscriptions;

    @Inject
    public PathTrackingPresenter(DataManagerDataTable dataManagerDataTable) {
        this.dataManagerDataTable = dataManagerDataTable;
        subscriptions = new CompositeSubscription();
    }

    @Override
    public void attachView(PathTrackingMvpView mvpView) {
        super.attachView(mvpView);
    }

    @Override
    public void detachView() {
        super.detachView();
        subscriptions.clear();
    }

    public void loadPathTracking(int userId) {
        checkViewAttached();
        getMvpView().showProgressbar(true);
        subscriptions.add(dataManagerDataTable.getUserPathTracking(userId)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe(new Subscriber<List<UserLocation>>() {
                    @Override
                    public void onCompleted() {

                    }

                    @Override
                    public void onError(Throwable e) {
                        getMvpView().showError();
                        getMvpView().showProgressbar(false);
                    }

                    @Override
                    public void onNext(List<UserLocation> userLocations) {
                        getMvpView().showProgressbar(false);
                        if (!userLocations.isEmpty()) {
                            getMvpView().showPathTracking(userLocations);
                        } else {
                            getMvpView().showEmptyPathTracking();
                        }
                    }
                })
        );
    }
}

File: mifosng-android/src/main/java/com/mifos/mifosxdroid/activity/pathtracking/PathTrackingPresenter.kt
package com.mifos.mifosxdroid.activity.pathtracking

import com.mifos.api.datamanager.DataManagerDataTable
import com.mifos.mifosxdroid.base.BasePresenter
import com.mifos.objects.user.UserLocation
import rx.Subscriber
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers
import rx.subscriptions.CompositeSubscription
import javax.inject.Inject

/**
 * Created by Rajan Maurya on 24/01/17.
 */
class PathTrackingPresenter @Inject constructor(private val dataManagerDataTable: DataManagerDataTable) :
    BasePresenter<PathTrackingMvpView>() {
    private val subscriptions: CompositeSubscription = CompositeSubscription()

    override fun attachView(mvpView: PathTrackingMvpView) {
        super.attachView(mvpView)
    }

    override fun detachView() {
        super.detachView()
        subscriptions.clear()
    }

    fun loadPathTracking(userId: Int) {
        checkViewAttached()
        mvpView?.showProgressbar(true)
        subscriptions.add(
            dataManagerDataTable.getUserPathTracking(userId)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe(object : Subscriber<List<UserLocation>>() {
                    override fun onCompleted() {}
                    override fun onError(e: Throwable) {
                        mvpView?.showError()
                        mvpView?.showProgressbar(false)
                    }

                    override fun onNext(userLocations: List<UserLocation>) {
                        mvpView?.showProgressbar(false)
                        if (userLocations.isNotEmpty()) {
                            mvpView?.showPathTracking(userLocations)
                        } else {
                            mvpView?.showEmptyPathTracking()
                        }
                    }
                })
        )
    }
}